packages/contracts/test/upgrade/dao.ts

```diff
@@ -0,0 +1,400 @@
+
+  context(`v1.0.0 to v1.3.0`, function () {
+    beforeEach(async function () {
+      daoV100Proxy = await deployWithProxy<DAO_V1_0_0>(DAO_V1_0_0);
+      await daoV100Proxy.initialize(
+        DUMMY_METADATA,
+        signers[0].address,
+        ethers.constants.AddressZero,
+        daoExampleURI
+      );
+
+      // Store the v1.0.0 implementation
+      daoV100Implementation = await readImplementationValueFromSlot(
+        daoV100Proxy.address
+      );
+
+      // Grant the upgrade permission
+      await daoV100Proxy.grant(
+        daoV100Proxy.address,
+        signers[0].address,
+        UPGRADE_PERMISSIONS.UPGRADE_DAO_PERMISSION_ID
+      );
+    });
+
+    it('does not corrupt the DAO storage', async () => {
+      // Upgrade and call `initializeFrom`.
+      const upgradeTx = await daoV100Proxy.upgradeToAndCall(
+        daoCurrentImplementaion.address,
+        DAO_Current.interface.encodeFunctionData('initializeFrom', [
+          [1, 0, 0],
+          EMPTY_DATA,
+        ])
+      );
+
+      // Check the stored implementation.
+      const implementationAfterUpgrade = await readImplementationValueFromSlot(
+        daoV100Proxy.address
+      );
+      expect(implementationAfterUpgrade).to.equal(
+        daoCurrentImplementaion.address
+      );
+      expect(implementationAfterUpgrade).to.not.equal(daoV100Implementation);
+
+      // Check the emitted implementation.
+      const emittedImplementation = (
+        await findEventTopicLog(upgradeTx, DAO_V1_0_0.interface, 'Upgraded')
+      ).args.implementation;
+      expect(emittedImplementation).to.equal(daoCurrentImplementaion.address);
+
+      // Check that storage is not corrupted.
+      expect(await daoV100Proxy.callStatic.daoURI()).to.equal(daoExampleURI);
+    });
+
+    it('does not corrupt permissions', async () => {
+      await daoV100Proxy.grant(
+        daoV100Proxy.address,
+        signers[0].address,
+        ethers.utils.id('EXECUTE_PERMISSION')
+      );
+      await daoV100Proxy.grant(
+        daoV100Proxy.address,
+        signers[0].address,
+        ethers.utils.id('ROOT_PERMISSION')
+      );
+
+      // Upgrade and call `initializeFrom`.
+      await daoV100Proxy.upgradeToAndCall(
+        daoCurrentImplementaion.address,
+        DAO_Current.interface.encodeFunctionData('initializeFrom', [
+          [1, 0, 0],
+          EMPTY_DATA,
+        ])
+      );
+
+      // Check the stored implementation.
+      const implementationAfterUpgrade = await readImplementationValueFromSlot(
+        daoV100Proxy.address
+      );
+      expect(implementationAfterUpgrade).to.equal(
+        daoCurrentImplementaion.address
+      );
+      expect(implementationAfterUpgrade).to.not.equal(daoV100Implementation);
+
+      expect(
+        await daoV100Proxy.hasPermission(
+          daoV100Proxy.address,
+          signers[0].address,
+          ethers.utils.id('EXECUTE_PERMISSION'),
+          EMPTY_DATA
+        )
+      ).to.be.true;
+      expect(
```

Comment:
Not really required, but I would also check that a permission that is not granted remains ungranted